File: VB/PersonalOrganizer/Common/DataModel/EntityFramework/DbReadOnlyRepository.vb
Imports System
Imports System.Linq
Imports System.Data.Entity
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Namespace Common.DataModel.EntityFramework
    Public MustInherit Class DbReadOnlyRepository(Of TEntity As Class, TDbContext As DbContext)
        Implements IReadOnlyRepository(Of TEntity)
        Private ReadOnly _dbSetAccessor As Func(Of TDbContext, DbSet(Of TEntity))
        Private ReadOnly _unitOfWork As DbUnitOfWork(Of TDbContext)
        Private _dbSet As DbSet(Of TEntity)
        Public Sub New(ByVal unitOfWork As DbUnitOfWork(Of TDbContext), ByVal dbSetAccessor As Func(Of TDbContext, DbSet(Of TEntity)))
            Me._dbSetAccessor = dbSetAccessor
            Me._unitOfWork = unitOfWork
        End Sub
        Protected ReadOnly Property DbSet As DbSet(Of TEntity)
            Get
                If _dbSet Is Nothing Then
                    _dbSet = _dbSetAccessor(_unitOfWork.Context)
                    'dbSet.Load();
                End If
                Return _dbSet
            End Get
        End Property
        Protected ReadOnly Property Context As TDbContext
            Get
                Return _unitOfWork.Context
            End Get
        End Property
        Protected Overridable Function GetEntities() As IQueryable(Of TEntity)
            Return DbSet
        End Function
        Private Function GetEntities_Impl() As IQueryable(Of TEntity) Implements IReadOnlyRepository(Of TEntity).GetEntities
            Return GetEntities()
        End Function
        Private ReadOnly Property UnitOfWork As IUnitOfWork Implements IReadOnlyRepository(Of TEntity).UnitOfWork
            Get
                Return _unitOfWork
            End Get
        End Property
        Private ReadOnly Property Local As ObservableCollection(Of TEntity) Implements IReadOnlyRepository(Of TEntity).Local
            Get
                Return DbSet.Local
            End Get
        End Property
    End Class
End Namespace

File: CS/PersonalOrganizer/Common/DataModel/EntityFramework/DbReadOnlyRepository.cs
using System;
using System.Linq;
using System.Data.Entity;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace PersonalOrganizer.Common.DataModel.EntityFramework {
    public abstract class DbReadOnlyRepository<TEntity, TDbContext> : IReadOnlyRepository<TEntity>
        where TEntity : class
        where TDbContext : DbContext {
        readonly Func<TDbContext, DbSet<TEntity>> dbSetAccessor;
        readonly DbUnitOfWork<TDbContext> unitOfWork;
        DbSet<TEntity> dbSet;

        public DbReadOnlyRepository(DbUnitOfWork<TDbContext> unitOfWork, Func<TDbContext, DbSet<TEntity>> dbSetAccessor) {
            this.dbSetAccessor = dbSetAccessor;
            this.unitOfWork = unitOfWork;
        }
        protected DbSet<TEntity> DbSet {
            get {
                if(dbSet == null) {
                    dbSet = dbSetAccessor(unitOfWork.Context);
                    //dbSet.Load();
                }
                return dbSet;
            }
        }
        protected TDbContext Context {
            get { return unitOfWork.Context; }
        }
        protected virtual IQueryable<TEntity> GetEntities() {
            return DbSet;
        }
        #region IReadOnlyRepository
        IQueryable<TEntity> IReadOnlyRepository<TEntity>.GetEntities() {
            return GetEntities();
        }
        IUnitOfWork IReadOnlyRepository<TEntity>.UnitOfWork {
            get { return unitOfWork; }
        }
        ObservableCollection<TEntity> IReadOnlyRepository<TEntity>.Local {
            get { return DbSet.Local; }
        }
        #endregion
    }
}
